lib/client/ipam.go

```diff
@@ -419,47 +423,58 @@ func (c ipams) assignFromExistingBlock(
 	return ips, nil
 }
 
-// ClaimAffinity claims affinity to the given host for all blocks
+// ClaimAffinity makes a best effort to claim affinity to the given host for all blocks
 // within the given CIDR.  The given CIDR must fall within a configured
-// pool.
-// TODO: Return indicators of claimed vs already claimed by another host.
-func (c ipams) ClaimAffinity(cidr common.IPNet, host *string) error {
+// pool.  Returns a list of blocks that were claimed, as well as a
+// list of blocks that were claimed by another host.
+func (c ipams) ClaimAffinity(cidr common.IPNet, host *string) ([]common.IPNet, []common.IPNet, error) {
 	// Validate that the given CIDR is at least as big as a block.
 	if !largerThanBlock(cidr) {
 		estr := fmt.Sprintf("The requested CIDR (%s) is smaller than the minimum.", cidr.String())
-		return invalidSizeError(estr)
+		return nil, nil, invalidSizeError(estr)
 	}
 
 	// Determine the hostname to use.
 	hostname := decideHostname(host)
+	failed := []common.IPNet{}
+	claimed := []common.IPNet{}
 
 	// Verify the requested CIDR falls within a configured pool.
 	if !c.blockReaderWriter.withinConfiguredPools(common.IP{cidr.IP}) {
 		estr := fmt.Sprintf("The requested CIDR (%s) is not within any configured pools.", cidr.String())
```

Comment:
Yeah, it does.

At the moment I'm not exposing different error types over the public API for IPAM.  That will be a part of a more comprehensive discussion on how we do error handling in libcalico (which we should have soon-ish).

Until then the errors are just treated opaquely.